### PR TITLE
Use warnings instead of print statements

### PR DESCRIPTION
**Pull request by StefanoFrazzetto**
#### Reference issue
Partially addresses: https://github.com/scipy/scipy/issues/1953

#### What does this implement/fix?
Use `warnings.warn` instead of print statements in [scipy/optimize/optimize.py](scipy/optimize/optimize.py), specifically within `def _minimize_bfgs` and `def _minimize_powell`

#### Additional information
Related PR with partial fixes for another function: https://github.com/scipy/scipy/pull/14050

As discussed with @WarrenWeckesser, it might be worth having a look at changes in `scipy/optimize/tests/test_optimize.py`, to understand if those tests always generate a warning and, if that's the case, use something like `with pytest.warns(OptimizeWarning):`, instead of just filtering the warnings.

### CODE REVIEW

File: scipy/optimize/tests/test_optimize.py [hunk 1]
```diff
@@ -1002,6 +1003,7 @@ def dfunc(z):
             assert_(func(sol1.x) < func(sol2.x),
                     "%s: %s vs. %s" % (method, func(sol1.x), func(sol2.x)))
 
+    @pytest.mark.filterwarnings('ignore::UserWarning')
```
Comment (StefanoFrazzetto) [inline scipy/optimize/tests/test_optimize.py:1006]: If the test always generates a warning, this should be replaced with `with pytest.warns(OptimizeWarning):`.

File: scipy/optimize/tests/test_optimize.py [hunk 2]
```diff
@@ -177,6 +177,7 @@ def test_bfgs(self):
                          [0, -5.24885582e-01, 4.87530347e-01]],
                         atol=1e-14, rtol=1e-7)
 
+    @pytest.mark.filterwarnings('ignore::UserWarning')
```
Comment (StefanoFrazzetto) [inline scipy/optimize/tests/test_optimize.py:180]: If the test always generates a warning, this should be replaced with `with pytest.warns(OptimizeWarning):`.

**Comment by StefanoFrazzetto** [comment]
I have tried the approach suggested by @mdhaber, i.e. creating a separate test with an expected failure, but passing a `maxiter` with a low value does not raise any warning. I run the tests a couple times, and I always get `Failed: DID NOT WARN. No warnings of type (<class 'scipy.optimize.optimize.OptimizeWarning'>,) was emitted`, using a `maxiter = 3`, or less.

**Comment by mdhaber** [comment]
Interesting. Can you tell how many iterations _are_ occurring when you give it `maxiter=3`? It's possible that the function is not obeying the iteration limit, and that would be another issue!

**Comment by mdhaber** [comment]
@StefanoFrazzetto would you be interested in continuing this if I were to help fix merge conflicts?

**Comment by StefanoFrazzetto** [comment]
@mdhaber it's been a while! If it adds any value to the project, I'll try to find some time and help you getting this merged. By the way, while setting up my env, I took the chance to resolve the conflicts, so hopefully this won't take too much effort. Let me know how I can help.

**Comment by mdhaber** [comment]
Hmm oops I didn't notice that gh-1953 was already resolved by gh-15259. This may not be needed anymore. 

The refactoring of the `minimize_powell` work is nice, but typically I wouldn't encourage refactoring unless it has some other added benefit. Perhaps we should have raised an `OptimizeWarning` from the start, but I'm not sure that we can now, since they are `RuntimeWarning`s, whereas `OptimizeWarning` is a subclass of `UserWarning`. If people are filtering them, their filters will break.

It looks like gh-15259 didn't change `bfgs`. 

@AtsushiSakai can I get your opinion here, since you worked on this before? What should this PR do now that the main issue (gh-1953) is resolved?

**Comment by AtsushiSakai** [comment]
I realized now that https://github.com/scipy/scipy/pull/15259 fixed almost bare print statements, but the `_minimize_bfgs` ones are missed:
https://github.com/scipy/scipy/blob/2f36c582363b6e75ea1fca68bd87043f4d0b983c/scipy/optimize/_optimize.py#L1441-L1445
So, I think it would be great that this PR fixs these.
I reopened https://github.com/scipy/scipy/issues/1953.

**Comment by mdhaber** [comment]
@AtsushiSakai what warning class should be used? You used `RuntimeWarning` for the others, but `OptimizeWarning` would be more appropriate.

**Comment by AtsushiSakai** [comment]
> but OptimizeWarning would be more appropriate.
Yeah. I didn't know the `OptimizeWarning`, but I agree it is more appropriate.

**Comment by mdhaber** [comment]
In that case, do you think we should replace the RuntimeWarnings with OptimizeWarnnings, then? It's slightly backward incompatible, but I think if an email were sent to the mailing list suggesting the change, it probably wouldn't bother people. 
If it does, then this PR could keep using them, and afterwards we could go back and change the others.
What do you think @StefanoFrazzetto @AtsushiSakai? Send an email?

**Comment by mdhaber** [comment]
@AtsushiSakai I was hoping to get your opinion on https://github.com/scipy/scipy/pull/14069#issuecomment-1273444118 when you have a moment.

File: scipy/optimize/tests/test_optimize.py (hunk 2 above)
Comment (mdhaber) [reply scipy/optimize/tests/test_optimize.py:212]: It doesn't.

**Comment by mdhaber** [comment]
Let's go ahead and leave these as `RuntimeWarnings` so we can merge this. @AtsushiSakai does this look good to you?

You can test locally with:
```python3
import numpy as np
from scipy import optimize

def f(x):
    return np.sum(x**2)

method = 'newton-cg'  # PR changed bfgs, cg, newton-cg, and powell
maxiter = 1000  # if 10, warning will be raised

options={'disp': True, 'maxiter': maxiter}
optimize.minimize(lambda x: optimize.rosen(x), [0, 0], method=method,
                  jac=optimize.rosen_der, options=options)
```

@StefanoFrazzetto would you be willing to add unit tests that try to raise the warning and check that it occurs?

**Review by AtsushiSakai — APPROVED** [review]
LGTM

**Comment by mdhaber** [comment]
OK, I fixed two little bugs and added tests. Since it was approved already, I'll go ahead and merge. Thanks, all!

**Pull request merged**